components/raftstore-v2/src/operation/command/admin/split.rs

```diff
@@ -0,0 +1,621 @@
+//!
+//! Execution:
+//! - exec_batch_split: Create and initialize metapb::region for split regions
+//!   and derived regions. Then, create checkpoints of the current talbet for
+//!   split regions and derived region to make tablet physical isolated. Update
+//!   the parent region's region state without persistency. Send the new regions
+//!   (including derived region) back to raftstore.
+//!
+//! Result apply:
+//! - todo
+//!
+//! Split peer creation and initlization:
+//! - todo
+//!
+//! Split finish:
+//! - todo
+
+use std::collections::VecDeque;
+
+use engine_traits::{
+    Checkpointer, KvEngine, OpenOptions, RaftEngine, TabletFactory, CF_DEFAULT, SPLIT_PREFIX,
+};
+use kvproto::{
+    metapb::Region,
+    raft_cmdpb::{AdminRequest, AdminResponse, RaftCmdRequest, SplitRequest},
+    raft_serverpb::RegionLocalState,
+};
+use protobuf::Message;
+use raftstore::{
+    coprocessor::split_observer::{is_valid_split_key, strip_timestamp_if_exists},
+    store::{
+        fsm::apply::validate_batch_split,
+        metrics::PEER_ADMIN_CMD_COUNTER,
+        util::{self, KeysInfoFormatter},
+        PeerStat, ProposalContext, RAFT_INIT_LOG_INDEX,
+    },
+    Result,
+};
+use slog::{info, warn, Logger};
+use tikv_util::box_err;
+
+use crate::{
+    batch::StoreContext,
+    fsm::ApplyResReporter,
+    operation::AdminCmdResult,
+    raft::{Apply, Peer},
+    router::ApplyRes,
+};
+
+#[derive(Debug)]
+pub struct SplitResult {
+    pub regions: Vec<Region>,
+    // The index of the derived region in `regions`
+    pub derived_index: usize,
+    pub tablet_index: u64,
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn propose_split<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        mut req: RaftCmdRequest,
+    ) -> Result<u64> {
+        validate_batch_split(req.mut_admin_request(), self.region())?;
+        let mut proposal_ctx = ProposalContext::empty();
+        proposal_ctx.insert(ProposalContext::SYNC_LOG);
+        proposal_ctx.insert(ProposalContext::SPLIT);
+
+        let data = req.write_to_bytes().unwrap();
+        self.propose_with_ctx(store_ctx, data, proposal_ctx.to_vec())
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    pub fn apply_split(
+        &mut self,
+        req: &AdminRequest,
+        log_index: u64,
+    ) -> Result<(AdminResponse, AdminCmdResult)> {
+        info!(
+            self.logger,
+            "split is deprecated, redirect to use batch split";
+        );
+        let split = req.get_split().to_owned();
+        let mut admin_req = AdminRequest::default();
+        admin_req
+            .mut_splits()
+            .set_right_derive(split.get_right_derive());
+        admin_req.mut_splits().mut_requests().push(split);
+        // This method is executed only when there are unapplied entries after being
+        // restarted. So there will be no callback, it's OK to return a response
+        // that does not matched with its request.
```

Comment:
I don't understand this comment. Where is the callback placed if there's one? Why response doesn't match the request in batch_split mode?

Reply:
It is copied from v1.